### ISSUE TITLE
AimsMIRegister does not work anymore in branch master

### ISSUE THREAD

**Issue by nsouedet** [brainvisa/aims-free#82]
On master branch, AimsMIRegister seg faults whereas it works well in 5.0 branch.

**To Reproduce**
1. Download attached file [aimsmiregister_issue.zip](https://github.com/brainvisa/aims-free/files/8001727/aimsmiregister_issue.zip)
2. Run AimsMIRegister -r mouse_atlas_brain_40micron_YZ_cutz_brd.ima -t Bh-template_3D_COR.ima
3. See the seg fault error

**Expected behavior**
The command should generate a transformation file named
Bh-template_3D_COR_TO_mouse_atlas_brain_40micron_YZ_cutz_brd.trm
that should contain:
0.00587783 -0.832984 1.82833
0.999961 -0.000171854 -0.00887041
0.00152378 0.988282 0.152629
0.00874024 -0.152637 0.988244

**Screenshots**
Not applicable

**Environment:**
 - Engine: Singularity
 - Version of BrainVISA: master branch
 - For local builds, version of the casa-dev image: casa-dev-5.0-8 (but it is not related to dev image)

**Additional context**
The issue was introduced by replacing AimsData with Volume in aims libraries.

**Comment by denisri** [comment]
Well, I could fix the crash and obtain a result, but it's not the same as the one you show. It doesn't look so bad (the registration is plausible), but it's different. As I don't think there is anything stochastic in the optimizer (right ?) and the older version of the command from the binary distribution brainvisa-5.0.4 produces the exact same result as you have, it seems that there is still something wrong in it.
I will commit the changes anyway, but it's not finished yet.

**Comment by denisri** [comment]
It was first broken in 31d9d2019e0d2ce76df02019368b258fe61ad04e (2021/08/24): from then `AimsMIRegister` did not crash but its output was a nearly identity matrix.
Then it started crashed, more recently.
Commit 31d9d2019e0d2ce76df02019368b258fe61ad04e was about moving `ChannelReader` to `Volume`.

**Comment by nsouedet** [comment]
Many thanks, Denis, for addressing this issue. You are so efficient ...
I don't know if it was an old problem that has been fixed by 31d9d2019e0d2ce76df02019368b258fe61ad04e. That is probably the case, because another idiot (me, of course) has coded this ChannelReader some years ago ...
Thanks again.

**Comment by nsouedet** [comment]
However, it might be interesting to understand why a ```ChannelReader<Volume>``` does not do the same thing as a ```ChannelReader<AimsData>```

**Comment by denisri** [comment]
That's what we have to look at ;)
However in the commit above, quite large portions of code have been removed (by me...) and I don't really know why, and it was possibly helpful somewhere...

**Comment by denisri** [comment]
I'm puzzled...
Starting from f28e55ee06592b52a1be915e847173f31d1d5640 (the commit just before it went wrong), the output from this commit is correct (same as the .trm in your "expected behaviour")
If I modify in AimsMIRegister, main.cc, lines 409-410:
```
    ChannelReader< AimsData<short> > rdref(fileRef);
    ChannelReader< AimsData<short> > rdtest(fileTest);
```
to:
```
    ChannelReader< VolumeRef<short> > rdref(fileRef);
    ChannelReader< VolumeRef<short> > rdtest(fileTest);
```
then I get incorrect results (and the process is twice as fast, lasting 14 minutes instead of 28).
The `ChannelReader` objects `rdref` and `rdtest` are each used at one single place, to read ref and test volumes.
If I print info about the volumes just after they are read (size, full header, contents of a few voxels at random locations), they are identical before and after the modification above. So it doesn't seem to change anything. And yet, this is what makes the difference...
I'll write the read images in order to compare them fully: there _must_ be something different in their reading...

**Comment by denisri** [comment]
`test` and `ref` are identical in both cases... (and identical to the input volumes) :/

**Comment by denisri** [comment]
Reference counters on the `Volume` are also the same.
I have no other idea, I'm stuck...

**Comment by denisri** [comment]
Now if we use:
```
    ChannelReader< VolumeRef<short> > rdref(fileRef);
    ChannelReader< AimsData<short> > rdtest(fileTest);
```
then the result is correct.
If we use, on the contrary:
```
    ChannelReader< AimsData<short> > rdref(fileRef);
    ChannelReader< VolumeRef<short> > rdtest(fileTest);
```
then the result is wrong. So it's in the test volume that something happens. But what ?

**Comment by denisri** [comment]
I think I'm beginning to understand... it's a weird side effect of casting non-const references to `Volume` from `AimsData`. I'm testing some things...

**Comment by denisri** [comment]
OK I get it. It's a more general flaw in `AimsData` / `Volume` cast.
When an `AimsData` is cast to a non-const `rc_ptr<Volume>`, we get the underlying `rc_ptr<Volume>` inside the `AimsData`. It seems good most of the time; but `AimsData` still holds a few variables that point to some `Volume` items: it has a pointer to its header (in an indirect form, through its `Header`/`PythonHeader` structure), and kind of strides in its `Border` structure.
If the cast `Volume` modifies significantly its structure: if it reallocates a new header, or changes size, then the origin `AimsData` variables are outdated, and nothing allows to update them. That's what happens in, typically, a `read()` function that reallocates and reads an existing `rc_ptr<Volume>`, such as in `ChannelReader<VolumeRef>::read( VolumeRef & )`, which is used in `AimsMIRegister` passing it an `AimsData` structure.
The good solution would be to get rid of all `AimsData` variables and make it directly use the `Volume`, but it's not so easy to change in an efficient way. I have to look at it...

**Comment by denisri** [comment]
I have done modifications in `AimsData` structure which could fix the first problem in 31d9d2019e0d2ce76df02019368b258fe61ad04e (2021/08/24), but it broke again in c9e4ed5b0c60ea6f1c2b69d764cf0634882e6be6 (2021/09/20) in a different way (the patch does not help here).
As I have not really tested my modifs (for #83) I have not pushed them yet. Anyway they don't fix things on the master branch.

**Comment by denisri** [comment]
There were still at least 2 other problems: one from taking address of an `AimsData` in MIRegister objective function, while it is passed a `rc_ptr<Volume>`, and the other is an index bug in `Sampler`. I'm testing fixes.

**Comment by denisri** [comment]
The last problem is in d42143fe92bfc92b55c240913841bfbda301f732, which actually changes the results. I need to completely reread it.

**Comment by denisri** [comment]
I've spent almost a week tracking this bug (or this series of bugs) through the history of changes, testing a series of patches applied to many repository commits, in a series of debug branches... Now I think I have fixed almost all "real bugs", and the difference in results appears to be a numerical precision difference:
In the older codes (`AimsData`) in `pdf.cc` in `AimsJointPVPdf()`, matrices are calculated, then get an intensity normalization: we process the sum of the elements, then divide all elements by this sum.
Using `AimsData`, the sum was calculated in a loop (`ForEach2d`) and summed in a float variable (same type as the matrix elements).
Using `Volume`, I have used `carto::sum(Volume)` or `Volume::sum()`, which extends the calculation type (during accumulation) to double. Even if the final sum is stored as a float, the result is different, just a bit, but this is enough to get a significantly different result. When displaying the registration result in Anatomist, we clearly see the difference, but both are "correctly registered" in the limits of the rigid registration possibilities. The former result (`AimsData`) seems a bit more "symmetric" to me than the newer (`Volume`), but I can't say it's really worse.
Now if we get the matrix sum as a double, we get still another result, which is still roughly valid.
The thing is that the registration is somewhat ambiguous, there are many matches which are similar in distance, thus the algorithm does not have a clear stable solution and is very sensible...

The other interesting thing is that, the more precise this sum, the faster the algorithm converges:
- sum as float (`AimsData`): the command runs for **27-28 minutes** on my laptop
- sum as double (`Volume`) but stored as float before division: **14-15 minutes**
- sum as double, stored as double: **11 minutes**.

As normally a better calculation precision leads to better results, and moreover here it is way faster, I'm about to adopt the double precision variant. Thus registration results will change compared to Aims 5.0.
There are several other places in `pdf.cc` where this kind of operation is done, I will change them also.
